network/quic/net.go

```diff
@@ -0,0 +1,160 @@
+package quic
+
+import (
+	"bufio"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"io"
+	"io/ioutil"
+	"log"
+	"math/big"
+	"sync"
+	"time"
+
+	h "github.com/ConsenSys/handel"
+	"github.com/ConsenSys/handel/network"
+	quic "github.com/lucas-clemente/quic-go"
+)
+
+// Network is a handel.Network implementation using QUIC as its transport layer
+type Network struct {
+	mutex          sync.RWMutex
+	listeners      []h.Listener
+	quit           bool
+	enc            network.Encoding
+	quicListener   quic.Listener
+	sessionManager sessionManager
+}
+
+const handshakeTimeout = 2000 * time.Millisecond
+
+// NewNetwork creates Nework baked by QUIC protocol
+func NewNetwork(addr string, enc network.Encoding) (*Network, error) {
+	cfg := generateTLSConfig()
+	qCfg := &quic.Config{HandshakeTimeout: handshakeTimeout} //, AcceptCookie: f}
+	listener, err := quic.ListenAddr(addr, cfg, qCfg)
+
+	if err != nil {
+		panic(err)
+	}
+	var listeners []h.Listener
+	sessManager := newSessionManager(handshakeTimeout)
+	net := Network{
+		mutex:          sync.RWMutex{},
+		listeners:      listeners,
+		quit:           false,
+		enc:            enc,
+		quicListener:   listener,
+		sessionManager: sessManager,
+	}
+
+	go net.handler()
+	return &net, nil
+}
+
+//RegisterListener registers listener for processing incoming packets
+func (quicNet *Network) RegisterListener(listener h.Listener) {
+	quicNet.mutex.Lock()
+	defer quicNet.mutex.Unlock()
+	quicNet.listeners = append(quicNet.listeners, listener)
+}
+
+// Stop stops the network
+func (quicNet *Network) Stop() {
+	quicNet.mutex.Lock()
+	defer quicNet.mutex.Unlock()
+	quicNet.quit = true
+}
+
+//Send sends a packet to supplied identities
+func (quicNet *Network) Send(identities []h.Identity, packet *h.Packet) {
+	for _, id := range identities {
+		go quicNet.send(id, packet)
```

Comment:
Mhhhhh... I'm not sure creating a go routine per packet is worth it, and might consume lots of ressources. Even if a go routine is small, handel intends to send quite lots of messages around. I see different potential solutions:
1. No go routine at all, the call is blocking and relies on the timeout of QUIC
2. One go routine per messages
3. One go routine per send-call - that's the lighter version compared to 2.
4. similar to 3 but one go routine that listens on channel and sends packets out, and each call to `send` just simply puts the packets in the channel - fifo style.

I'm fine with 1, 3 and 4, although 1 should be sufficient at the moment and pretty fairly easily done ^^ but let's discuss that.

Reply:
we need to be careful, if just one node is not available then "send" to that node will block all the other calls. For now the handshake is is set to 1s. this means for one "absent" node the loop will freeze for 1s (and it will get worse with more offline nodes), but I agree one go routine per packet might be too much